### ISSUE TITLE
[java] CompareObjectsWithEquals - false negative with type res

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#2880]
**Affects PMD Version:** 6.30.0-SNAPSHOT

**Rule:** [CompareObjectsWithEquals](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#compareobjectswithequals)

**Description:**

From https://github.com/pmd/pmd/pull/2871#issuecomment-717848659

**Code Sample demonstrating the issue:**

See https://chunk.io/pmd/fc7db65b2c6a46eca4c9a0c3012482e2/diff/checkstyle/index.html#A111

```java
import java.math.BigInteger;

public class Foo {
    private BigInteger field = new BigInteger("10");
    private BigInteger field2 = new BigInteger("20");
    
    public boolean check(final BigInteger param) {
        final BigInteger other = new BigInteger("20");

        return (field == null
                || param != field   // warning expected: Use equals() to compare object references. 
                && other != field)  // warning expected: Use equals() to compare object references. 
            && param.testBit(field2.intValue())
            && other.testBit(field2.intValue());
    }
}
```

**Expected outcome:**

*   Is PMD missing to report a violation, where there should be one? -> false-negative

**Running PMD through:** regression-tester

**Comment by adangel** [comment]
Fixed via #2899 for PMD 7.
